File: 01. Elevator/ElevatorSystem.java
import java.util.*;

enum Direction {
    UP, DOWN, IDLE
}

class Elevator {
    int currentFloor;
    Direction direction;
    boolean isMoving;

    public Elevator() {
        this.currentFloor = 0;
        this.direction = Direction.IDLE;
        this.isMoving = false;
    }

    public void moveUp() {
        currentFloor++;
        direction = Direction.UP;
        isMoving = true;
        System.out.println("Elevator moving up to floor: " + currentFloor);
    }

    public void moveDown() {
        currentFloor--;
        direction = Direction.DOWN;
        isMoving = true;
        System.out.println("Elevator moving down to floor: " + currentFloor);
    }

    public void stop() {
        direction = Direction.IDLE;
        isMoving = false;
        System.out.println("Elevator stopped at floor: " + currentFloor + "\n");
    }

    public boolean shouldMoveUp(int floor) {
        return currentFloor < floor;
    }

    public boolean shouldMoveDown(int floor) {
        return currentFloor > floor;
    }
}

class ElevatorController {
    Elevator elevator;
    PriorityQueue<Integer> upRequests;
    PriorityQueue<Integer> downRequests;
    Set<Integer> processedRequests;
    Map<Integer, Integer> destinationFloors;

    public ElevatorController() {
        this.elevator = new Elevator();
        this.upRequests = new PriorityQueue<>(); // Min-heap for UP direction
        this.downRequests = new PriorityQueue<>(Collections.reverseOrder()); // Max-heap for DOWN direction
        this.processedRequests = new HashSet<>();
        this.destinationFloors = new HashMap<>();
    }

    public void addRequest(int floor, Direction direction, int destinationFloor) {
        if (processedRequests.contains(floor)) return; // Skip if already processed
        processedRequests.add(floor);
        destinationFloors.put(floor, destinationFloor);

        if (direction == Direction.UP) {
            upRequests.offer(floor);
        } else if (direction == Direction.DOWN) {
            downRequests.offer(floor);
        }
    }

    public void processRequests() {
        while (!upRequests.isEmpty() || !downRequests.isEmpty()) {
            if (elevator.direction == Direction.IDLE) {
                determineInitialDirection();
            }

            if (elevator.direction == Direction.UP) {
                handleUpRequests();
                if (upRequests.isEmpty() && !downRequests.isEmpty()) {
                    elevator.direction = Direction.DOWN;
                }
            } else if (elevator.direction == Direction.DOWN) {
                handleDownRequests();
                if (downRequests.isEmpty() && !upRequests.isEmpty()) {
                    elevator.direction = Direction.UP;
                }
            }
        }
    }

    private void determineInitialDirection() {
        elevator.direction = Direction.UP;
    }

    private void handleUpRequests() {
        while (!upRequests.isEmpty() && elevator.direction == Direction.UP) {
            int nextFloor = upRequests.poll();
            moveToFloor(nextFloor);
        }
    }

    private void handleDownRequests() {
        while (!downRequests.isEmpty() && elevator.direction == Direction.DOWN) {
            int nextFloor = downRequests.poll();
            moveToFloor(nextFloor);
        }
    }

    private void moveToFloor(int floor) {
        while (elevator.currentFloor != floor) {
            if (elevator.shouldMoveUp(floor)) {
                elevator.moveUp();
            } else if (elevator.shouldMoveDown(floor)) {
                elevator.moveDown();
            }
        }
        elevator.stop();
        addRequest(destinationFloors.get(floor), 
                  (destinationFloors.get(floor) - floor) > 0 ? Direction.UP : Direction.DOWN, 
                  destinationFloors.get(floor));
        processedRequests.remove(floor);
    }
}

public class ElevatorSystem {
    public static void main(String[] args) {
        ElevatorController controller = new ElevatorController();

        // Simulate requests
        controller.addRequest(3, Direction.UP, 7);
        controller.addRequest(1, Direction.DOWN, 0);
        controller.addRequest(5, Direction.UP, 6);
        controller.addRequest(2, Direction.DOWN, 1);

        // Process all requests
        controller.processRequests();
    }
}

File: Low Level Design (LLD)/Problems/03. ElevatorSystem/Cpp/ElevatorSystem.cpp
#include <bits/stdc++.h>
using namespace std;

enum Direction { UP, DOWN, IDLE };

class Elevator {
public:
    int currentFloor;
    Direction direction;
    bool isMoving;

    Elevator() {
        currentFloor = 0;
        direction = IDLE;
        isMoving = false;
    }

    void moveUp() {
        currentFloor++;
        direction = UP;
        isMoving = true;
        cout << "Elevator moving up to floor: " << currentFloor << endl;
    }

    void moveDown() {
        currentFloor--;
        direction = DOWN;
        isMoving = true;
        cout << "Elevator moving down to floor: " << currentFloor << endl;
    }

    void stop() { 
        direction = IDLE;
        isMoving = false;
        cout << "Elevator stopped at floor: " << currentFloor << endl << endl;
    }

    bool shouldMoveUp(int floor) {
        return currentFloor < floor;
    }

    bool shouldMoveDown(int floor) {
        return currentFloor > floor;
    }
};

class ElevatorController {
public:
    Elevator elevator;
    
    priority_queue<int, vector<int>, greater<int>> upRequests; // Min-heap for UP direction
    priority_queue<int> downRequests; // Max-heap for DOWN direction
    unordered_set<int> processedRequests; // Track processed requests
    unordered_map<int, int> destinationFloors;

    void addRequest(int floor, Direction direction, int destinationFloor) {
        if (processedRequests.find(floor) != processedRequests.end()) return; // Skip if already processed
        processedRequests.insert(floor);
        destinationFloors[floor] = destinationFloor;

        if (direction == UP) {
            upRequests.push(floor);
        }
        else if (direction == DOWN) {
            downRequests.push(floor);
        }
    }

    void processRequests() {
        while (!upRequests.empty() || !downRequests.empty()) {
            if (elevator.direction == IDLE) {
                determineInitialDirection();
            }

            if (elevator.direction == UP) {
                handleUpRequests();
                if (upRequests.empty() && !downRequests.empty()) {
                    elevator.direction = DOWN;
                }
            }
            else if (elevator.direction == DOWN) {
                handleDownRequests();
                if (downRequests.empty() && !upRequests.empty()) {
                    elevator.direction = UP;
                }
            }
        }
    }

private:
    void determineInitialDirection() {
        elevator.direction = UP;
    };

    void handleUpRequests() {
        while (!upRequests.empty() && elevator.direction == UP) {
            int nextFloor = upRequests.top();
            upRequests.pop();
            moveToFloor(nextFloor);
        }
    };
    
    void handleDownRequests() {
        while (!downRequests.empty() && elevator.direction == DOWN) {
            int nextFloor = downRequests.top();
            downRequests.pop();
            moveToFloor(nextFloor);
        }
    }

    void moveToFloor(int floor) {
        while (elevator.currentFloor != floor) {
            if (elevator.shouldMoveUp(floor)) {
                elevator.moveUp();
            }
            else if (elevator.shouldMoveDown(floor)) {
                elevator.moveDown();
            }
        }
        elevator.stop();
        this->addRequest(destinationFloors[floor], (destinationFloors[floor] - floor) > 0 ? UP : DOWN, destinationFloors[floor]);
        processedRequests.erase(floor);
    }
};

int main() {
    ElevatorController controller;

    // Simulate requests
    controller.addRequest(3, UP, 7);
    controller.addRequest(1, DOWN, 0);
    controller.addRequest(5, UP, 6);
    controller.addRequest(2, DOWN, 1);

    // Process all requests
    controller.processRequests();

    return 0;
}